posthog/hogql/printer.py

```diff
@@ -468,7 +469,17 @@ def visit_call(self, node: ast.Call):
                 )
 
             if self.dialect == "clickhouse":
-                if node.name == "concat":
+                if node.name.startswith("tumble") or node.name.startswith("hop"):
+                    args: List[str] = []
+                    for idx, arg in enumerate(node.args):
+                        if idx == 0:
+                            if arg.name in ADD_OR_NULL_DATETIME_FUNCTIONS:
+                                args.append(f"assumeNotNull(toDateTime({self.visit(arg)}))")
+                            else:
+                                args.append(f"toDateTime({self.visit(arg)})")
```

Comment:
I see more and more hints that we need to work towards having full support for types in HogQL. This code will work if the call happens directly, but not if you do something like

```sql
select tumble(x.c, interval 1 day) from events left join (select c as toDateTime(now())) x ...
```

A type system would help catch this, but nothing really to do now.

I think one change needed though, as we can't be sure `arg` is a function call:

```suggestion
                            if isinstance(arg, ast.Call) and arg.name in ADD_OR_NULL_DATETIME_FUNCTIONS:
                                args.append(f"assumeNotNull(toDateTime({self.visit(arg)}))")
                            else:
                                args.append(f"toDateTime({self.visit(arg)})")
```